wallet/invoices.c

```diff
@@ -0,0 +1,402 @@
+	stmt = db_prepare(invoices->db, "DELETE FROM invoices WHERE id=?;");
+	sqlite3_bind_int64(stmt, 1, invoice->id);
+	db_exec_prepared(invoices->db, stmt);
+
+	if (sqlite3_changes(invoices->db->sql) != 1)
+		return false;
+
+	/* Delete from invoices object. */
+	list_del_from(&invoices->invlist, &invoice->list);
+
+	/* Tell all the waiters about the fact that it was deleted. */
+	while ((w = list_pop(&invoice->waitone_waiters,
+			     struct invoice_waiter,
+			     list)) != NULL) {
+		/* Acquire the watcher for ourself first. */
+		tal_steal(tmpctx, w);
+		trigger_invoice_waiter(w, NULL);
+	}
+
+	/* Free all watchers and the invoice. */
+	tal_free(tmpctx);
+	tal_free(invoice);
+	return true;
+}
+
+const struct invoice *invoices_iterate(struct invoices *invoices,
+				       const struct invoice *invoice)
+{
+	if (invoice)
+		return list_next(&invoices->invlist, invoice, list);
+	else
+		return list_top(&invoices->invlist, struct invoice, list);
+}
+
+static s64 get_next_pay_index(struct db *db)
+{
+	/* Equivalent to (next_pay_index++) */
+	s64 next_pay_index;
+	next_pay_index = db_get_intvar(db, "next_pay_index", 0);
+	/* Variable should exist. */
+	assert(next_pay_index > 0);
+	db_set_intvar(db, "next_pay_index", next_pay_index + 1);
+	return next_pay_index;
+}
+
+
+void invoices_resolve(struct invoices *invoices,
+		      const struct invoice *cinvoice,
+		      u64 msatoshi_received)
+{
+	sqlite3_stmt *stmt;
+	struct invoice_waiter *w;
+	struct invoice *invoice = (struct invoice *)cinvoice;
+	s64 pay_index;
+	const tal_t *tmpctx = tal_tmpctx(NULL);
+
+	/* Assign a pay-index. */
+	pay_index = get_next_pay_index(invoices->db);
+	/* FIXME: Save time of payment. */
+
+	/* Update database. */
+	stmt = db_prepare(invoices->db,
+			  "UPDATE invoices"
+			  "   SET state=?"
+			  "     , pay_index=?"
+			  "     , msatoshi_received=?"
+			  " WHERE id=?;");
+	sqlite3_bind_int(stmt, 1, PAID);
+	sqlite3_bind_int64(stmt, 2, pay_index);
+	sqlite3_bind_int64(stmt, 3, msatoshi_received);
+	sqlite3_bind_int64(stmt, 4, invoice->id);
+	db_exec_prepared(invoices->db, stmt);
+
+	/* Update in-memory structure. */
+	invoice->state = PAID;
+	invoice->pay_index = pay_index;
+	invoice->msatoshi_received = msatoshi_received;
+
+	/* Tell all the waitany waiters about the new paid invoice. */
+	while ((w = list_pop(&invoices->waitany_waiters,
+			     struct invoice_waiter,
+			     list)) != NULL) {
+		tal_steal(tmpctx, w);
+		trigger_invoice_waiter(w, invoice);
+	}
+	/* Tell any waitinvoice waiters about the specific invoice
+	 * getting paid. */
+	while ((w = list_pop(&invoice->waitone_waiters,
+			     struct invoice_waiter,
+			     list)) != NULL) {
+		tal_steal(tmpctx, w);
+		trigger_invoice_waiter(w, invoice);
+	}
+
+	/* Free all watchers. */
+	tal_free(tmpctx);
+}
+
+/* Called when an invoice waiter is destructed. */
+static void invoice_waiter_dtor(struct invoice_waiter *w)
+{
+	/* Already triggered. */
+	if (w->triggered)
+		return;
+	list_del(&w->list);
```

Comment:
Could this ever not delete the invoice from `w->list`? `trigger_invoice_waiter` doesn't remove it.

Reply:
The `invoices_resolve` function pops waiters from the `waitany_waiters` and `waitone_waiters` lists prior to triggering.  The `list_pop` deletes the invoice from those lists.  Note that `list_del` warns that when a node is deleted from a list, it can only be `list_add*`-ed again, and in particular that it should not be `list_del`ed again.

Notice that in `invoices_resolve` we steal waiter objects into the `tmpctx` after we pop the waiter from the list.  So the destructor will get called just before `invoices_resolve` returns, and the object is no longer in the list (it was popped off) and in particular should not call `list_del` on the waiter object again.

Reply:
Ok, sounds good.